File: backend/sentry.py
#!/usr/bin/env python3
"""
Project Sentry - Automated Workspace Audit System
Monitors /home/a2/Desktop/gem for changes and generates daily reports
"""

import os
import json
import subprocess
import shutil
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Tuple

WORKSPACE = Path("/home/a2/Desktop/gem")
STATE_FILE = Path("/home/a2/Desktop/gem/opb/backend/.sentry_state.json")
IGNORE_PATTERNS = {'.git', '__pycache__', 'node_modules', '.venv', 'browser_data'}

class SentryAudit:
    def __init__(self):
        self.workspace = WORKSPACE
        self.state_file = STATE_FILE
        self.current_snapshot = {}
        self.previous_snapshot = {}
        self.report = []
        
    def load_previous_state(self):
        """Load the last snapshot from disk"""
        if self.state_file.exists():
            try:
                with open(self.state_file, 'r') as f:
                    data = json.load(f)
                    self.previous_snapshot = data.get('files', {})
                    return data.get('timestamp')
            except: pass
        return None
    
    def save_current_state(self):
        """Persist current snapshot for next run"""
        state = {
            'timestamp': datetime.now().isoformat(),
            'files': self.current_snapshot
        }
        with open(self.state_file, 'w') as f:
            json.dump(state, f, indent=2)
    
    def scan_filesystem(self):
        """Walk workspace and capture file metadata"""
        for root, dirs, files in os.walk(self.workspace):
            # Filter ignored directories
            dirs[:] = [d for d in dirs if d not in IGNORE_PATTERNS]
            
            for file in files:
                filepath = Path(root) / file
                try:
                    stat = filepath.stat()
                    rel_path = str(filepath.relative_to(self.workspace))
                    self.current_snapshot[rel_path] = {
                        'size': stat.st_size,
                        'mtime': int(stat.st_mtime)
                    }
                except: pass
    
    def analyze_changes(self) -> Dict:
        """Compare current vs previous snapshot"""
        current_files = set(self.current_snapshot.keys())
        previous_files = set(self.previous_snapshot.keys())
        
        created = current_files - previous_files
        deleted = previous_files - current_files
        modified = []
        
        for path in current_files & previous_files:
            if self.current_snapshot[path] != self.previous_snapshot.get(path):
                modified.append(path)
        
        # Calculate size changes
        total_delta = 0
        for path in created:
            total_delta += self.current_snapshot[path]['size']
        for path in deleted:
            total_delta -= self.previous_snapshot[path]['size']
        
        return {
            'created': list(created),
            'modified': modified,
            'deleted': list(deleted),
            'total_delta_mb': total_delta / (1024 * 1024),
            'churn': len(created) + len(modified) + len(deleted)
        }
    
    def get_git_status(self) -> Dict:
        """Query git repository status"""
        if not (self.workspace / '.git').exists():
            return {'available': False}
        
        try:
            # Current branch
            branch = subprocess.check_output(
                ['git', 'branch', '--show-current'],
                cwd=self.workspace, text=True
            ).strip()
            
            # Porcelain status
            status = subprocess.check_output(
                ['git', 'status', '--porcelain=v1'],
                cwd=self.workspace, text=True
            ).strip()
            
            lines = status.split('\n') if status else []
            modified = sum(1 for l in lines if l.startswith(' M'))
            staged = sum(1 for l in lines if l.startswith('M '))
            untracked = sum(1 for l in lines if l.startswith('??'))
            
            # Last commit
            try:
                last_commit = subprocess.check_output(
                    ['git', 'log', '-1', '--format=%h|%cr'],
                    cwd=self.workspace, text=True
                ).strip().split('|')
                commit_hash = last_commit[0]
                commit_age = last_commit[1]
            except:
                commit_hash = 'N/A'
                commit_age = 'N/A'
            
            return {
                'available': True,
                'branch': branch,
                'clean': len(lines) == 0,
                'modified': modified,
                'staged': staged,
                'untracked': untracked,
                'last_commit': commit_hash,
                'commit_age': commit_age
            }
        except:
            return {'available': False, 'error': 'Failed to query git'}
    
    def get_disk_usage(self) -> Dict:
        """Check disk and workspace usage"""
        usage = shutil.disk_usage(self.workspace)
        
        # Workspace size
        try:
            result = subprocess.check_output(
                ['du', '-sb', str(self.workspace)],
                text=True
            )
            workspace_bytes = int(result.split()[0])
        except:
            workspace_bytes = 0
        
        return {
            'total_gb': usage.total / (1024**3),
            'used_gb': usage.used / (1024**3),
            'free_gb': usage.free / (1024**3),
            'usage_percent': (usage.used / usage.total) * 100,
            'workspace_gb': workspace_bytes / (1024**3)
        }
    
    def generate_report(self, changes: Dict, git: Dict, disk: Dict) -> str:
        """Generate markdown report"""
        lines = []
        lines.append("=== PROJECT SENTRY DAILY REPORT ===")
        lines.append(f"Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        lines.append(f"Workspace: {self.workspace}")
        lines.append("")
        
        # Summary
        lines.append("[SUMMARY]")
        churn_level = "HIGH" if changes['churn'] > 20 else "MEDIUM" if changes['churn'] > 5 else "LOW"
        lines.append(f"File changes: +{len(changes['created'])} ~{len(changes['modified'])} -{len(changes['deleted'])}   (Churn: {churn_level})")
        
        if git['available']:
            git_status = "CLEAN" if git['clean'] else "DIRTY"
            lines.append(f"Git status: {git_status} (branch: {git['branch']})")
        
        disk_status = "WARNING" if disk['usage_percent'] > 80 else "OK"
        lines.append(f"Disk usage: {disk['usage_percent']:.0f}% ({disk_status})")
        lines.append("")
        
        # Filesystem
        lines.append("[FILESYSTEM]")
        lines.append(f"Created: {len(changes['created'])}")
        lines.append(f"Modified: {len(changes['modified'])}")
        lines.append(f"Deleted: {len(changes['deleted'])}")
        lines.append(f"Size delta: {changes['total_delta_mb']:+.1f} MB")
        
        if changes['created'][:3]:
            lines.append("Top created:")
            for f in changes['created'][:3]:
                lines.append(f"  - {f}")
        lines.append("")
        
        # Git
        if git['available']:
            lines.append("[GIT]")
            lines.append(f"Branch: {git['branch']}")
            lines.append(f"Modified: {git['modified']}")
            lines.append(f"Staged: {git['staged']}")
            lines.append(f"Untracked: {git['untracked']}")
            lines.append(f"Last commit: {git['commit_age']} ({git['last_commit']})")
            lines.append("")
        
        # Resources
        lines.append("[RESOURCES]")
        lines.append(f"Disk: {disk['total_gb']:.0f}G total / {disk['free_gb']:.0f}G free ({disk['usage_percent']:.0f}%)")
        lines.append(f"Workspace size: {disk['workspace_gb']:.1f}G")
        lines.append("")
        
        # Alerts
        alerts = []
        if git['available'] and not git['clean']:
            alerts.append("Uncommitted changes present")
        if disk['usage_percent'] > 80:
            alerts.append("Disk usage above 80%")
        if changes['churn'] > 50:
            alerts.append("High file churn detected")
        
        if alerts:
            lines.append("[ALERTS]")
            for alert in alerts:
                lines.append(f"- {alert}")
            lines.append("")
        
        # Actions
        actions = []
        if git['available'] and not git['clean']:
            actions.append("Commit or stash working tree")
        if disk['usage_percent'] > 80:
            actions.append("Review disk usage and clean up old files")
        
        if actions:
            lines.append("[RECOMMENDED ACTIONS]")
            for action in actions:
                lines.append(f"- {action}")
            lines.append("")
        
        lines.append("Roger. Over.")
        return '\n'.join(lines)
    
    def run(self) -> str:
        """Execute full audit cycle"""
        last_run = self.load_previous_state()
        self.scan_filesystem()
        
        changes = self.analyze_changes()
        git = self.get_git_status()
        disk = self.get_disk_usage()
        
        report = self.generate_report(changes, git, disk)
        
        self.save_current_state()
        return report

import time
import sys

if __name__ == "__main__":
    sentry = SentryAudit()
    
    # Check for --once flag for single-run mode (used by !audit command)
    if "--once" in sys.argv:
        try:
            report = sentry.run()
            print(report, flush=True)
        except Exception as e:
            print(f"Error in audit: {e}", flush=True)
            sys.exit(1)
    else:
        # Daemon mode - continuous monitoring
        print("Starting Sentry Audit Daemon...", flush=True)
        while True:
            try:
                report = sentry.run()
                print(report, flush=True)
            except Exception as e:
                print(f"Error in audit cycle: {e}", flush=True)
            
            # Run every 60 seconds
            time.sleep(60)

File: backend/src/sentry.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::PathBuf;
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};
use serde::{Deserialize, Serialize};
use walkdir::WalkDir;
use chrono::Local;


const WORKSPACE: &str = "/home/a2/Desktop/gem";
const STATE_FILE: &str = "/home/a2/Desktop/gem/opb/backend/.sentry_state.json";
const IGNORE_PATTERNS: &[&str] = &[".git", "__pycache__", "node_modules", ".venv", "browser_data"];

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct FileMetadata {
    pub size: u64,
    pub mtime: u64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SentryState {
    pub timestamp: String,
    pub files: HashMap<String, FileMetadata>,
}

pub struct SentryAudit {
    workspace: PathBuf,
    state_file: PathBuf,
    current_snapshot: HashMap<String, FileMetadata>,
    previous_snapshot: HashMap<String, FileMetadata>,
}

impl SentryAudit {
    pub fn new() -> Self {
        Self {
            workspace: PathBuf::from(WORKSPACE),
            state_file: PathBuf::from(STATE_FILE),
            current_snapshot: HashMap::new(),
            previous_snapshot: HashMap::new(),
        }
    }

    pub fn load_previous_state(&mut self) {
        if let Ok(content) = fs::read_to_string(&self.state_file) {
            if let Ok(state) = serde_json::from_str::<SentryState>(&content) {
                self.previous_snapshot = state.files;
            }
        }
    }

    pub fn save_current_state(&self) {
        let state = SentryState {
            timestamp: Local::now().to_rfc3339(),
            files: self.current_snapshot.clone(),
        };
        if let Ok(json) = serde_json::to_string_pretty(&state) {
            let _ = fs::write(&self.state_file, json);
        }
    }

    pub fn scan_filesystem(&mut self) {
        let walker = WalkDir::new(&self.workspace).into_iter();
        for entry in walker.filter_entry(|e| {
            !IGNORE_PATTERNS.iter().any(|p| e.file_name().to_string_lossy() == *p)
        }) {
            if let Ok(entry) = entry {
                if entry.file_type().is_file() {
                    if let Ok(metadata) = entry.metadata() {
                        let rel_path = entry.path().strip_prefix(&self.workspace)
                            .unwrap_or(entry.path())
                            .to_string_lossy().to_string();
                        
                        let mtime = metadata.modified().unwrap_or(SystemTime::now())
                            .duration_since(UNIX_EPOCH).unwrap_or_default().as_secs();

                        self.current_snapshot.insert(rel_path, FileMetadata {
                            size: metadata.len(),
                            mtime,
                        });
                    }
                }
            }
        }
    }

    pub fn analyze_changes(&self) -> serde_json::Value {
        let current_keys: HashSet<_> = self.current_snapshot.keys().collect();
        let previous_keys: HashSet<_> = self.previous_snapshot.keys().collect();

        let created: Vec<_> = current_keys.difference(&previous_keys).map(|k| k.to_string()).collect();
        let deleted: Vec<_> = previous_keys.difference(&current_keys).map(|k| k.to_string()).collect();
        let mut modified = Vec::new();

        for key in current_keys.intersection(&previous_keys) {
            if self.current_snapshot.get(*key) != self.previous_snapshot.get(*key) {
                modified.push(key.to_string());
            }
        }

        let mut total_delta_bytes: i64 = 0;
        for path in &created {
            total_delta_bytes += self.current_snapshot[path].size as i64;
        }
        for path in &deleted {
            total_delta_bytes -= self.previous_snapshot[path].size as i64;
        }

        serde_json::json!({
            "created": created,
            "modified": modified,
            "deleted": deleted,
            "total_delta_mb": total_delta_bytes as f64 / (1024.0 * 1024.0),
            "churn": created.len() + modified.len() + deleted.len()
        })
    }

    pub fn get_git_status(&self) -> serde_json::Value {
        if !self.workspace.join(".git").exists() {
            return serde_json::json!({"available": false});
        }

        let branch = Command::new("git").arg("branch").arg("--show-current").current_dir(&self.workspace).output().ok()
            .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string()).unwrap_or_default();

        let status_out = Command::new("git").arg("status").arg("--porcelain=v1").current_dir(&self.workspace).output().ok()
            .map(|o| String::from_utf8_lossy(&o.stdout).to_string()).unwrap_or_default();

        let lines: Vec<&str> = status_out.lines().collect();
        let modified = lines.iter().filter(|l| l.starts_with(" M")).count();
        let staged = lines.iter().filter(|l| l.starts_with("M ")).count();
        let untracked = lines.iter().filter(|l| l.starts_with("??")).count();

        let last_commit = Command::new("git").arg("log").arg("-1").arg("--format=%h|%cr").current_dir(&self.workspace).output().ok()
            .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string()).unwrap_or_default();
        let commit_parts: Vec<&str> = last_commit.split('|').collect();

        serde_json::json!({
            "available": true,
            "branch": branch,
            "clean": lines.is_empty(),
            "modified": modified,
            "staged": staged,
            "untracked": untracked,
            "last_commit": commit_parts.get(0).unwrap_or(&"N/A"),
            "commit_age": commit_parts.get(1).unwrap_or(&"N/A")
        })
    }

    pub fn get_disk_usage(&self) -> serde_json::Value {
        use sysinfo::Disks;
        let disks = Disks::new_with_refreshed_list();
        
        let mut total_b: u64 = 1;
        let mut avail_b: u64 = 0;
        
        for disk in &disks {
            total_b = disk.total_space();
            avail_b = disk.available_space();
            break; // Use the first disk as in Python shutil.disk_usage
        }
        let used_b = total_b.saturating_sub(avail_b);

        let workspace_bytes = Command::new("du").arg("-sb").arg(&self.workspace).output().ok()
            .and_then(|o| String::from_utf8_lossy(&o.stdout).split_whitespace().next().map(|s| s.parse::<u64>().unwrap_or(0)))
            .unwrap_or(0);

        serde_json::json!({
            "total_gb": total_b / (1024 * 1024 * 1024),
            "used_gb": used_b / (1024 * 1024 * 1024),
            "free_gb": avail_b / (1024 * 1024 * 1024),
            "usage_percent": (used_b as f64 / total_b as f64) * 100.0,
            "workspace_gb": workspace_bytes as f64 / (1024.0 * 1024.0 * 1024.0)
        })
    }

    pub fn generate_report(&self, changes: &serde_json::Value, git: &serde_json::Value, disk: &serde_json::Value) -> String {
        let mut lines = Vec::new();
        lines.push("=== PROJECT SENTRY DAILY REPORT ===".to_string());
        lines.push(format!("Date: {}", Local::now().format("%Y-%m-%d %H:%M:%S")));
        lines.push(format!("Workspace: {}", self.workspace.display()));
        lines.push("".to_string());

        lines.push("[SUMMARY]".to_string());
        let churn = changes["churn"].as_u64().unwrap_or(0);
        let churn_level = if churn > 20 { "HIGH" } else if churn > 5 { "MEDIUM" } else { "LOW" };
        lines.push(format!("File changes: +{} ~{} -{}   (Churn: {})", 
            changes["created"].as_array().map(|a| a.len()).unwrap_or(0),
            changes["modified"].as_array().map(|a| a.len()).unwrap_or(0),
            changes["deleted"].as_array().map(|a| a.len()).unwrap_or(0),
            churn_level));

        if git["available"].as_bool().unwrap_or(false) {
            let status = if git["clean"].as_bool().unwrap_or(false) { "CLEAN" } else { "DIRTY" };
            lines.push(format!("Git status: {} (branch: {})", status, git["branch"].as_str().unwrap_or("")));
        }

        let usage_percent = disk["usage_percent"].as_f64().unwrap_or(0.0);
        let disk_status = if usage_percent > 80.0 { "WARNING" } else { "OK" };
        lines.push(format!("Disk usage: {:.0}% ({})", usage_percent, disk_status));
        lines.append(&mut vec!["".to_string(), "[FILESYSTEM]".to_string()]);
        
        lines.push(format!("Created: {}", changes["created"].as_array().map(|a| a.len()).unwrap_or(0)));
        lines.push(format!("Modified: {}", changes["modified"].as_array().map(|a| a.len()).unwrap_or(0)));
        lines.push(format!("Deleted: {}", changes["deleted"].as_array().map(|a| a.len()).unwrap_or(0)));
        lines.push(format!("Size delta: {:+.1} MB", changes["total_delta_mb"].as_f64().unwrap_or(0.0)));

        if let Some(created) = changes["created"].as_array() {
            if !created.is_empty() {
                lines.push("Top created:".to_string());
                for f in created.iter().take(3) {
                    lines.push(format!("  - {}", f.as_str().unwrap_or("")));
                }
            }
        }
        lines.push("".to_string());

        if git["available"].as_bool().unwrap_or(false) {
            lines.push("[GIT]".to_string());
            lines.push(format!("Branch: {}", git["branch"].as_str().unwrap_or("")));
            lines.push(format!("Modified: {}", git["modified"]));
            lines.push(format!("Staged: {}", git["staged"]));
            lines.push(format!("Untracked: {}", git["untracked"]));
            lines.push(format!("Last commit: {} ({})", git["commit_age"].as_str().unwrap_or(""), git["last_commit"].as_str().unwrap_or("")));
            lines.push("".to_string());
        }

        lines.push("[RESOURCES]".to_string());
        lines.push(format!("Disk: {}G total / {}G free ({:.0}%)", 
            disk["total_gb"], disk["free_gb"], usage_percent));
        lines.push(format!("Workspace size: {:.1}G", disk["workspace_gb"].as_f64().unwrap_or(0.0)));
        lines.push("".to_string());

        let mut alerts = Vec::new();
        if git["available"].as_bool().unwrap_or(false) && !git["clean"].as_bool().unwrap_or(false) {
            alerts.push("Uncommitted changes present");
        }
        if usage_percent > 80.0 {
            alerts.push("Disk usage above 80%");
        }
        if churn > 50 {
            alerts.push("High file churn detected");
        }

        if !alerts.is_empty() {
            lines.push("[ALERTS]".to_string());
            for alert in alerts {
                lines.push(format!("- {}", alert));
            }
            lines.push("".to_string());
        }

        let mut actions = Vec::new();
        if git["available"].as_bool().unwrap_or(false) && !git["clean"].as_bool().unwrap_or(false) {
            actions.push("Commit or stash working tree");
        }
        if usage_percent > 80.0 {
            actions.push("Review disk usage and clean up old files");
        }

        if !actions.is_empty() {
            lines.push("[RECOMMENDED ACTIONS]".to_string());
            for action in actions {
                lines.push(format!("- {}", action));
            }
            lines.push("".to_string());
        }

        lines.push("Roger. Over.".to_string());
        lines.join("\n")
    }

    pub fn run(&mut self) -> String {
        self.load_previous_state();
        self.scan_filesystem();
        let changes = self.analyze_changes();
        let git = self.get_git_status();
        let disk = self.get_disk_usage();
        let report = self.generate_report(&changes, &git, &disk);
        self.save_current_state();
        report
    }
}
